src/utils/object.js

```diff
@@ -44,3 +46,54 @@ export function map(opt_initial) {
 export function hasOwn(obj, key) {
   return hasOwn_.call(obj, key);
 }
+
+/**
+ * @param {!Object} target
+ * @param {!Object} source
+ * @param {number} currentDepth
+ * @param {number} maxDepth The maximum depth for deep merge, beyond which
+ *    Object.assign will be used.
+ * @return {!Object}
+ */
+function deepMerge_(target, source, currentDepth, maxDepth) {
+  if (currentDepth > maxDepth) {
+    Object.assign(target, source);
+    return target;
+  }
+  Object.keys(source).forEach(key => {
+    const newValue = source[key];
+    // Perform a deep merge IFF both a and b have the same property and
+    // the properties on both a and b are non-null plain objects
+    if (hasOwn(target, key)) {
+      const oldValue = target[key];
+      if (isObject(newValue) && isObject(oldValue)) {
+        target[key] = deepMerge_(
+            oldValue,
+            newValue,
+            currentDepth + 1,
+            maxDepth);
+        return;
+      }
+    }
+    target[key] = newValue;
+  });
+  return target;
+}
+
+/**
+ * Deep merge object b into object a. Both a and b can only contain
+ * primitives, arrays, and plain objects. For any conflicts, object b wins.
+ * Arrays are replaced, not merged. Plain objects are recursively merged.
+ * @param {!Object} target
+ * @param {!Object} source
+ * @param {number|undefined=} opt_maxDepth The maximum depth for deep merge,
```

Comment:
`{number=}`

BTW you were right that this is `undefined` if not provided which I believe behaves like `NaN` in comparison ops.